### PR TITLE
WSIReader defaults and tensor conversion

### LINKED ISSUE #6056: `WSIReader` defaults at initialization
Many WSIReader defaults are being set at `get_data()` method but it would be useful to be able to set those default values to be set in `__init__` so the WSIReader with defaults can be passed as a reader for wsi datasets and inference.
For instance, this is a required feature for whole slide patch inference to enable users to pass arguments for wsireader to the splitter.

### PR DESCRIPTION
**Pull request by drbeh**
Fixes #6056 

### Description

This PR enable to instantiate `WSIReader` classes with default values for `mode` and `dtype`. It also accepts `torch.dtype` as dtype and if `torch.dtype` is provided, the output will be a `torch.Tensor` to avoid additional step to convert the output for inference.

### Types of changes

- [x] Non-breaking change (fix or new feature that would not break existing functionality).
- [x] New tests added to cover the changes.
- [x] Integration tests passed locally by running `./runtests.sh -f -u --net --coverage`.
- [x] Quick tests passed locally by running `./runtests.sh --quick --unittests  --disttests`.
- [x] In-line docstrings updated.

### CODE REVIEW

File: monai/data/wsi_reader.py [hunk 1]
```diff
@@ -34,12 +35,19 @@ class BaseWSIReader(ImageReader):
     """
     An abstract class that defines APIs to load patches from whole slide image files.
 
+    Args:
+        level: the whole slide image level at which the image is extracted.
+        channel_dim: the desired dimension for color channel.
+        dtype: the data type of output image.
+        mode: the output image color mode, e.g., "RGB" or "RGBA".
+        kwargs: additional args for the reader
```
Comment (KumoLiu) [inline monai/data/wsi_reader.py:43]: Seems missing the `device` doc-string here.

File: monai/data/wsi_reader.py [hunk 2]
```diff
@@ -281,21 +316,40 @@ class WSIReader(BaseWSIReader):
         backend: the name of backend whole slide image reader library, the default is cuCIM.
         level: the level at which patches are extracted.
         channel_dim: the desired dimension for color channel. Default to 0 (channel first).
+        dtype: the data type of output image. Defaults to `np.uint8`.
+        mode: the output image color mode, "RGB" or "RGBA". Defaults to "RGB".
```
Comment (KumoLiu) [inline monai/data/wsi_reader.py:320]: Also missing the doc-string for `device` here.

File: monai/data/wsi_reader.py [hunk 3]
```diff
@@ -403,6 +457,8 @@ class CuCIMWSIReader(BaseWSIReader):
         level: the whole slide image level at which the image is extracted. (default=0)
             This is overridden if the level argument is provided in `get_data`.
         channel_dim: the desired dimension for color channel. Default to 0 (channel first).
+        dtype: the data type of output image. Defaults to `np.uint8`.
+        mode: the output image color mode, "RGB" or "RGBA". Defaults to "RGB".
```
Comment (KumoLiu) [inline monai/data/wsi_reader.py:461]: The same comments as before.

File: monai/data/wsi_reader.py [hunk 4]
```diff
@@ -551,13 +616,26 @@ class OpenSlideWSIReader(BaseWSIReader):
...
```
Comment (KumoLiu) [inline monai/data/wsi_reader.py:620]: Same here.

**Review by Nic-Ma — APPROVED** [review]

**Comment by Nic-Ma** [comment]
Hi @drbeh ,

I think the error is as below:
```
2023-03-08T02:13:10.5286648Z [2023-03-08T02:12:42.239Z] ======================================================================
2023-03-08T02:13:10.5287076Z [2023-03-08T02:12:42.239Z] FAIL: test_read_region_device_1__home_jenkins_agent_workspace_MONAI_premerge_monai_tests_testing_data_temp_CMU_1_tiff_tiff (tests.test_wsireader.TestCuCIM)
2023-03-08T02:13:10.5287345Z [2023-03-08T02:12:42.239Z] ----------------------------------------------------------------------
2023-03-08T02:13:10.5287547Z [2023-03-08T02:12:42.239Z] Traceback (most recent call last):
2023-03-08T02:13:10.5287895Z [2023-03-08T02:12:42.239Z]   File "/opt/conda/lib/python3.8/site-packages/parameterized/parameterized.py", line 533, in standalone_func
2023-03-08T02:13:10.5288106Z [2023-03-08T02:12:42.239Z]     return func(*(a + p.args), **p.kwargs)
2023-03-08T02:13:10.5288467Z [2023-03-08T02:12:42.239Z]   File "/home/jenkins/agent/workspace/MONAI-premerge/monai/tests/test_wsireader.py", line 562, in test_read_region_device
2023-03-08T02:13:10.5288715Z [2023-03-08T02:12:42.239Z]     self.assertEqual(img.device, torch.device(device))
2023-03-08T02:13:10.5289093Z [2023-03-08T02:12:42.239Z] AssertionError: device(type='cuda', index=0) != device(type='cuda')
2023-03-08T02:13:10.5289226Z [2023-03-08T02:12:42.239Z] 
2023-03-08T02:13:10.5289452Z [2023-03-08T02:12:42.239Z] ======================================================================
2023-03-08T02:13:10.5289868Z [2023-03-08T02:12:42.239Z] FAIL: test_read_region_device_3__home_jenkins_agent_workspace_MONAI_premerge_monai_tests_testing_data_temp_CMU_1_tiff_tiff (tests.test_wsireader.TestCuCIM)
2023-03-08T02:13:10.5290134Z [2023-03-08T02:12:42.239Z] ----------------------------------------------------------------------
2023-03-08T02:13:10.5290336Z [2023-03-08T02:12:42.239Z] Traceback (most recent call last):
2023-03-08T02:13:10.5290683Z [2023-03-08T02:12:42.239Z]   File "/opt/conda/lib/python3.8/site-packages/parameterized/parameterized.py", line 533, in standalone_func
2023-03-08T02:13:10.5290898Z [2023-03-08T02:12:42.239Z]     return func(*(a + p.args), **p.kwargs)
2023-03-08T02:13:10.5291260Z [2023-03-08T02:12:42.239Z]   File "/home/jenkins/agent/workspace/MONAI-premerge/monai/tests/test_wsireader.py", line 559, in test_read_region_device
2023-03-08T02:13:10.5291465Z [2023-03-08T02:12:42.239Z]     assert_allclose(img, expected_img)
2023-03-08T02:13:10.5291790Z [2023-03-08T02:12:42.239Z]   File "/home/jenkins/agent/workspace/MONAI-premerge/monai/tests/utils.py", line 124, in assert_allclose
2023-03-08T02:13:10.5292123Z [2023-03-08T02:12:42.239Z]     np.testing.assert_equal(isinstance(actual, np.ndarray), isinstance(desired, np.ndarray), "numpy type")
2023-03-08T02:13:10.5292448Z [2023-03-08T02:12:42.239Z]   File "/opt/conda/lib/python3.8/site-packages/numpy/testing/_private/utils.py", line 425, in assert_equal
2023-03-08T02:13:10.5292642Z [2023-03-08T02:12:42.239Z]     raise AssertionError(msg)
2023-03-08T02:13:10.5292811Z [2023-03-08T02:12:42.239Z] AssertionError: 
2023-03-08T02:13:10.5293009Z [2023-03-08T02:12:42.239Z] Items are not equal: numpy type
2023-03-08T02:13:10.5293226Z [2023-03-08T02:12:42.239Z]  ACTUAL: True
2023-03-08T02:13:10.5293378Z [2023-03-08T02:12:42.239Z]  DESIRED: False
2023-03-08T02:13:10.5293508Z [2023-03-08T02:12:42.239Z] 
2023-03-08T02:13:10.5293734Z [2023-03-08T02:12:42.239Z] ======================================================================
2023-03-08T02:13:10.5294158Z [2023-03-08T02:12:42.239Z] FAIL: test_read_region_device_5__home_jenkins_agent_workspace_MONAI_premerge_monai_tests_testing_data_temp_CMU_1_tiff_tiff (tests.test_wsireader.TestCuCIM)
2023-03-08T02:13:10.5294426Z [2023-03-08T02:12:42.239Z] ----------------------------------------------------------------------
2023-03-08T02:13:10.5294627Z [2023-03-08T02:12:42.239Z] Traceback (most recent call last):
2023-03-08T02:13:10.5294973Z [2023-03-08T02:12:42.239Z]   File "/opt/conda/lib/python3.8/site-packages/parameterized/parameterized.py", line 533, in standalone_func
2023-03-08T02:13:10.5295222Z [2023-03-08T02:12:42.239Z]     return func(*(a + p.args), **p.kwargs)
2023-03-08T02:13:10.5295583Z [2023-03-08T02:12:42.239Z]   File "/home/jenkins/agent/workspace/MONAI-premerge/monai/tests/test_wsireader.py", line 559, in test_read_region_device
2023-03-08T02:13:10.5295788Z [2023-03-08T02:12:42.239Z]     assert_allclose(img, expected_img)
2023-03-08T02:13:10.5296115Z [2023-03-08T02:12:42.239Z]   File "/home/jenkins/agent/workspace/MONAI-premerge/monai/tests/utils.py", line 124, in assert_allclose
2023-03-08T02:13:10.5296447Z [2023-03-08T02:12:42.239Z]     np.testing.assert_equal(isinstance(actual, np.ndarray), isinstance(desired, np.ndarray), "numpy type")
2023-03-08T02:13:10.5296772Z [2023-03-08T02:12:42.239Z]   File "/opt/conda/lib/python3.8/site-packages/numpy/testing/_private/utils.py", line 425, in assert_equal
2023-03-08T02:13:10.5296954Z [2023-03-08T02:12:42.239Z]     raise AssertionError(msg)
2023-03-08T02:13:10.5297121Z [2023-03-08T02:12:42.239Z] AssertionError: 
2023-03-08T02:13:10.5297316Z [2023-03-08T02:12:42.239Z] Items are not equal: numpy type
2023-03-08T02:13:10.5297469Z [2023-03-08T02:12:42.239Z]  ACTUAL: True
2023-03-08T02:13:10.5297632Z [2023-03-08T02:12:42.239Z]  DESIRED: False
2023-03-08T02:13:10.5297764Z [2023-03-08T02:12:42.239Z] 
2023-03-08T02:13:10.5297991Z [2023-03-08T02:12:42.239Z] ======================================================================
2023-03-08T02:13:10.5298418Z [2023-03-08T02:12:42.239Z] FAIL: test_read_region_device_1__home_jenkins_agent_workspace_MONAI_premerge_monai_tests_testing_data_temp_CMU_1_tiff_tiff (tests.test_wsireader.TestTiffFile)
2023-03-08T02:13:10.5298686Z [2023-03-08T02:12:42.239Z] ----------------------------------------------------------------------
2023-03-08T02:13:10.5298884Z [2023-03-08T02:12:42.239Z] Traceback (most recent call last):
2023-03-08T02:13:10.5299224Z [2023-03-08T02:12:42.239Z]   File "/opt/conda/lib/python3.8/site-packages/parameterized/parameterized.py", line 533, in standalone_func
2023-03-08T02:13:10.5299437Z [2023-03-08T02:12:42.239Z]     return func(*(a + p.args), **p.kwargs)
2023-03-08T02:13:10.5299787Z [2023-03-08T02:12:42.239Z]   File "/home/jenkins/agent/workspace/MONAI-premerge/monai/tests/test_wsireader.py", line 562, in test_read_region_device
2023-03-08T02:13:10.5300030Z [2023-03-08T02:12:42.239Z]     self.assertEqual(img.device, torch.device(device))
2023-03-08T02:13:10.5300291Z [2023-03-08T02:12:42.239Z] AssertionError: device(type='cuda', index=0) != device(type='cuda')
2023-03-08T02:13:10.5300423Z [2023-03-08T02:12:42.239Z] 
2023-03-08T02:13:10.5300648Z [2023-03-08T02:12:42.239Z] ======================================================================
2023-03-08T02:13:10.5301075Z [2023-03-08T02:12:42.239Z] FAIL: test_read_region_device_3__home_jenkins_agent_workspace_MONAI_premerge_monai_tests_testing_data_temp_CMU_1_tiff_tiff (tests.test_wsireader.TestTiffFile)
2023-03-08T02:13:10.5301337Z [2023-03-08T02:12:42.239Z] ----------------------------------------------------------------------
2023-03-08T02:13:10.5301540Z [2023-03-08T02:12:42.239Z] Traceback (most recent call last):
2023-03-08T02:13:10.5301928Z [2023-03-08T02:12:42.239Z]   File "/opt/conda/lib/python3.8/site-packages/parameterized/parameterized.py", line 533, in standalone_func
2023-03-08T02:13:10.5302136Z [2023-03-08T02:12:42.239Z]     return func(*(a + p.args), **p.kwargs)
2023-03-08T02:13:10.5302494Z [2023-03-08T02:12:42.239Z]   File "/home/jenkins/agent/workspace/MONAI-premerge/monai/tests/test_wsireader.py", line 559, in test_read_region_device
2023-03-08T02:13:10.5302698Z [2023-03-08T02:12:42.239Z]     assert_allclose(img, expected_img)
2023-03-08T02:13:10.5303022Z [2023-03-08T02:12:42.239Z]   File "/home/jenkins/agent/workspace/MONAI-premerge/monai/tests/utils.py", line 124, in assert_allclose
2023-03-08T02:13:10.5303354Z [2023-03-08T02:12:42.239Z]     np.testing.assert_equal(isinstance(actual, np.ndarray), isinstance(desired, np.ndarray), "numpy type")
2023-03-08T02:13:10.5303711Z [2023-03-08T02:12:42.239Z]   File "/opt/conda/lib/python3.8/site-packages/numpy/testing/_private/utils.py", line 425, in assert_equal
2023-03-08T02:13:10.5303910Z [2023-03-08T02:12:42.239Z]     raise AssertionError(msg)
2023-03-08T02:13:10.5304076Z [2023-03-08T02:12:42.239Z] AssertionError: 
2023-03-08T02:13:10.5304273Z [2023-03-08T02:12:42.239Z] Items are not equal: numpy type
2023-03-08T02:13:10.5304428Z [2023-03-08T02:12:42.239Z]  ACTUAL: True
2023-03-08T02:13:10.5304588Z [2023-03-08T02:12:42.239Z]  DESIRED: False
2023-03-08T02:13:10.5304718Z [2023-03-08T02:12:42.239Z] 
2023-03-08T02:13:10.5304942Z [2023-03-08T02:12:42.239Z] ======================================================================
2023-03-08T02:13:10.5305369Z [2023-03-08T02:12:42.239Z] FAIL: test_read_region_device_5__home_jenkins_agent_workspace_MONAI_premerge_monai_tests_testing_data_temp_CMU_1_tiff_tiff (tests.test_wsireader.TestTiffFile)
2023-03-08T02:13:10.5305632Z [2023-03-08T02:12:42.239Z] ----------------------------------------------------------------------
2023-03-08T02:13:10.5305832Z [2023-03-08T02:12:42.239Z] Traceback (most recent call last):
2023-03-08T02:13:10.5306178Z [2023-03-08T02:12:42.239Z]   File "/opt/conda/lib/python3.8/site-packages/parameterized/parameterized.py", line 533, in standalone_func
2023-03-08T02:13:10.5306382Z [2023-03-08T02:12:42.239Z]     return func(*(a + p.args), **p.kwargs)
2023-03-08T02:13:10.5306737Z [2023-03-08T02:12:42.239Z]   File "/home/jenkins/agent/workspace/MONAI-premerge/monai/tests/test_wsireader.py", line 559, in test_read_region_device
2023-03-08T02:13:10.5306941Z [2023-03-08T02:12:42.239Z]     assert_allclose(img, expected_img)
2023-03-08T02:13:10.5307272Z [2023-03-08T02:12:42.239Z]   File "/home/jenkins/agent/workspace/MONAI-premerge/monai/tests/utils.py", line 124, in assert_allclose
2023-03-08T02:13:10.5307602Z [2023-03-08T02:12:42.239Z]     np.testing.assert_equal(isinstance(actual, np.ndarray), isinstance(desired, np.ndarray), "numpy type")
2023-03-08T02:13:10.5307925Z [2023-03-08T02:12:42.239Z]   File "/opt/conda/lib/python3.8/site-packages/numpy/testing/_private/utils.py", line 425, in assert_equal
2023-03-08T02:13:10.5308117Z [2023-03-08T02:12:42.239Z]     raise AssertionError(msg)
2023-03-08T02:13:10.5308286Z [2023-03-08T02:12:42.239Z] AssertionError: 
2023-03-08T02:13:10.5308477Z [2023-03-08T02:12:42.239Z] Items are not equal: numpy type
2023-03-08T02:13:10.5308632Z [2023-03-08T02:12:42.239Z]  ACTUAL: True
2023-03-08T02:13:10.5308791Z [2023-03-08T02:12:42.239Z]  DESIRED: False
2023-03-08T02:13:10.5308922Z [2023-03-08T02:12:42.239Z] 
2023-03-08T02:13:10.5309182Z [2023-03-08T02:12:42.239Z] ----------------------------------------------------------------------
2023-03-08T02:13:10.5309366Z [2023-03-08T02:12:42.239Z] Ran 12302 tests in 1393.784s
2023-03-08T02:13:10.5309496Z [2023-03-08T02:12:42.239Z] 
2023-03-08T02:13:10.5309693Z [2023-03-08T02:12:42.239Z] FAILED (failures=6, skipped=521)
```

Thanks.

**Comment by drbeh** [comment]
@Nic-Ma it should work now.

**Comment by drbeh** [comment]
@Nic-Ma I am making few changes. Please hold on a bit. Thanks

**Comment by drbeh** [comment]
@Nic-Ma it is ready, and has passed all the tests locally. Thanks

**Comment by drbeh** [comment]
@Nic-Ma @wyli it seems that the issue with blossom has fixed. Can we run blossom tests again here?

**Comment by drbeh** [comment]
@Nic-Ma @wyli @ericspod do you have any idea why `quick-py3 (ubuntu-latest)` are getting cancelled?
https://github.com/Project-MONAI/MONAI/actions/runs/4406004944/jobs/7717649759#step:12:17631

The only thing that I can think of is memory limit since TiffFile loads the entire image at level 0 for `test_read_region_1__Users_runner_work_MONAI_MONAI_tests_testing_data_temp_CMU_1_tiff_tiff (tests.test_wsireader.TestTiffFile)`, where the test always gets cancelled.

However, the same test are going well on Mac `quick-py3 (macOS-latest)`:
https://github.com/Project-MONAI/MONAI/actions/runs/4406004944/jobs/7717649413#step:12:17755

Do we have different memory limits on the instances that Mac and Ubuntu quick tests are running?

File: monai/data/wsi_reader.py [hunk 5]
```diff
@@ -58,13 +75,41 @@ class BaseWSIReader(ImageReader):
...
+
+    def set_device(self, device):
+        if device is None:
+            self.device = None
+        elif isinstance(device, torch.device):
+            self.device = device
+        elif isinstance(device, str):
+            self.device = torch.device(device)
+        else:
+            raise ValueError(f"`device` must be `torch.device`, `str` or `None` but {type(device)} is given.")
```
[the remark this answers is not in the document]
Comment (drbeh) [reply monai/data/wsi_reader.py:111]: You are right! I know that but here I wanted to make sure that `self.device` of `WSIReader` is always a valid `torch.device` so we can use `self.device.type` for comparing to string without checking again if it is string or torech.device.
https://github.com/Project-MONAI/MONAI/blob/0d0d6c1420bd472f0ab62eb6d25bed5332184b6d/monai/data/wsi_reader.py#L278

[the remark this answers is not in the document]
Comment (wyli) [reply monai/data/wsi_reader.py:111]: you can still convert without a isinstance check:
```
>>> import torch
>>> a = torch.device("cpu")
>>> torch.device(a)
device(type='cpu')
```

also this excludes any possible extensions based on other cuda arrays such as cupy implementation, I don't think it's good for a base class...

[the remark this answers is not in the document]
Comment (drbeh) [reply monai/data/wsi_reader.py:111]: The logic is if output is np.ndarray and device is cuda, I want to create a torch tensor, but if it is cpu, don't, so I don't see how it is possible without comparison.

[the remark this answers is not in the document]
Comment (drbeh) [reply monai/data/wsi_reader.py:111]: I see! you mean conversion only for comparison. Sure, make sense.

**Comment by wyli** [comment]
it seems the memory issue is still here https://github.com/Project-MONAI/MONAI/actions/runs/4418529796/jobs/7745780000 could you please have a look soon @drbeh it might block 1.2rc releases... cc @Nic-Ma @KumoLiu

**Pull request merged**